sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala

```diff
@@ -66,27 +75,22 @@ trait BaseLimitExec extends UnaryExecNode with CodegenSupport {
   // to the parent operator.
   override def usedInputs: AttributeSet = AttributeSet.empty
 
+  private lazy val countTerm = BaseLimitExec.newLimitCountTerm()
+
+  override lazy val conditionsOfKeepProducingData: Seq[String] = {
+    s"$countTerm < $limit" +: super.conditionsOfKeepProducingData
```

Comment:
note that this is sub-optimal for adjacent limits, but I think it's fine as optimizer will merge adjacent limits.